rr_graph_uxsdcxx.h

```diff
@@ -0,0 +1,398 @@
+#endif
+
+
+/* Stores a vector of elements in a shared pool. Consists
+ * of an offset into the pool and a size. It's faster than a regular
+ * vector, but one can only insert into it when it's the last vector
+ * in the pool. */
+template<class T, std::vector<T> &pool>
+class collapsed_vec {
+private:
+	uint32_t _size;
+	uint32_t _offset;
+public:
+	inline collapsed_vec(){
+		_size = 0;
+		_offset = pool.size();
+	}
+	inline T& back(){
+		return pool[_offset+_size-1];
+	}
+	inline T* begin(){
+		return &pool[_offset];
+	}
+	inline T* end(){
+		return &pool[_offset+_size];
+	}
+	inline T& operator[](uint32_t i){
+		return pool[_offset+i];
+	}
+	inline void push_back(const T &x){
+		assert(_size+_offset == pool.size());
+		pool.push_back(x);
+		_size++;
+	}
+	inline uint32_t size(){
+		return _size;
+	}
+};
+
+/* A pool of strings. It exposes an add(const char *x) function
+ * which is like a better strdup.
+ * add() calls strlen on the argument to see how much space is needed.
+ * Then it checks its vector of chunks to see if there is a chunk it
+ * would fit into. If none is found, a new one is allocated with a size of
+ * max(length of string, last chunk's size*2). */
+class string_pool_impl {
+private:
+	const uint32_t INITIAL_SIZE = 1024;
+	struct chunk {
+		uint32_t used;
+		uint32_t size;
+		char * mem;
+	};
+	std::vector<chunk> chunks;
+public:
+	string_pool_impl(){
+		chunk c;
+		c.used = 0;
+		c.size = INITIAL_SIZE;
+		c.mem = (char *)std::malloc(INITIAL_SIZE);
+		chunks.emplace_back(c);
+	}
+	~string_pool_impl(){
+		for(auto &c: chunks)
+			free(c.mem);
+	}
+	const char *add(const char *x){
+		uint32_t len = std::strlen(x)+1;
+		char *out;
+		for(auto &c: chunks){
+			if(c.used+len <= c.size){
+				out = &c.mem[c.used];
+				c.used += len;
+				std::memcpy(out, x, len);
+				return out;
+			}
+		}
+		chunk n;
+		n.used = 0;
+		n.size = std::max(chunks.back().size*2, len);
+		n.mem = (char *)std::malloc(n.size);
+		std::memcpy(n.mem, x, len);
+		chunks.emplace_back(n);
+		return n.mem;
+	}
+	void clear(){
+		for(int i=1; i<chunks.size(); i++){
+			free(chunks[i].mem);
+		}
+		chunks.resize(1);
+		chunks[0].used = 0;
+		std::memset(chunks[0].mem, 0, INITIAL_SIZE);
+	}
+};
+/* Forward decl of generated data types. Needed for the pools. */
+struct t_channel;
+struct t_x_list;
+struct t_y_list;
+struct t_timing;
+struct t_sizing;
+struct t_segment_timing;
+struct t_pin;
+struct t_meta;
+struct t_grid_loc;
+struct t_node_loc;
+struct t_node_timing;
+struct t_node_segment;
+struct t_bin_nodes;
+struct t_connection_box;
+struct t_switch;
+struct t_segment;
+struct t_pin_class;
+struct t_metadata;
+struct t_channels;
+struct t_grid_locs;
+struct t_connection_boxes;
+struct t_block_type;
+struct t_node;
+struct t_edge;
+struct t_switches;
+struct t_segments;
+struct t_block_types;
+struct t_rr_nodes;
+struct t_rr_edges;
+struct t_rr_graph;
+
+/* Global shared pools for storing multiply-occurring elements. */
```

Comment:
How are these sorted?

Reply:
Arbitrary- depends on the order of `anno_element` calls: https://github.com/duck2/uxsdcxx/blob/master/uxsdcxx.py#L182 I should define an order.